Userland/chmod.cpp

```diff
@@ -25,3 +60,94 @@ int main(int argc, char** argv)
 
     return 0;
 }
+
+int str_to_octal_mode(char access_scope, char** access_str, unsigned& mode)
+{
+    char op_type = **access_str;
+    int rc = 0;
+
+    if(op_type != '+' && op_type != '-' && op_type != '=') {
+        return -1;
+    }
+    if(op_type == '=') {
+        switch (access_scope) {
+            case 'u':
+                mode &= ~(S_IRUSR | S_IWUSR | S_IXUSR); break;
+            case 'g':
+                mode &= ~(S_IRGRP | S_IWGRP | S_IXGRP); break;
+            case 'o':
+                mode &= ~(S_IROTH | S_IWOTH | S_IXOTH); break;
+            case 'a':
+                mode &= ~(S_IRUSR | S_IWUSR | S_IXUSR |
+                          S_IRGRP | S_IWGRP | S_IXGRP |
+                          S_IROTH | S_IWOTH | S_IXOTH); break;
+        }
+	    op_type = '+';
+    }
+    (*access_str)++;
+    while(**access_str != '\0' && **access_str != ',') {
+        rc = add_perm(access_scope, **access_str, op_type, mode);
+        if(rc != 0) {
+            return -1;
+        }
+        (*access_str)++;
+    }
+
+    return 0;
+}
+
+int add_perm(char access_scope, char access_perm, char op_type, unsigned& mode)
+{
+    if(access_perm != 'r' && access_perm != 'w' && access_perm != 'x') {
+        return -1;
+    }
+
+    switch(access_scope) {
+        case 'u':
+            switch(access_perm) {
+                case 'r':
+                    (op_type == '+' ? mode |= S_IRUSR : mode &= (~S_IRUSR));
```

Comment:
This type of control flow is overly obfuscating for no reason.
Instead, let's just compute a `mask` in the big switch table, e.g:
```cpp
case 'r':
    mask = S_IRUSR;
    break;
```
Then at the very end, you can apply the mask like so:
```cpp
if (op_type == '+')
    mode |= mask;
else
    mode &= ~mask;
```